### PR TITLE
cmd/validate-krew-manifest: initial implementation

### LINKED ISSUE #148: hack:CI/CD tooling to verify manifests
We should have a way of doing some sanity checks on the plugin files to prevent:

- the entire index from breaking due to bad yaml
- bad links/sha (at the time of merging the PR)

Perhaps we can find the changed yaml files and do `krew install --source=$f` to verify installation.

### PR DESCRIPTION
**Pull request by ahmetb**
This tool takes a manifest file and runs certain validation steps. It's meant
to test plugin manifests added to [krew-index](https://github.com/GoogleContainerTools/krew-index) install correctly (so far it has
found one broken manifest).

Fixes #148.
Related to #149 but doesn't fix it.

### CODE REVIEW

**Comment by ahmetb** [comment]
Fixes #148.
Fixes #149.

**Comment by ahmetb** [comment]
cc: @juanvallejo 
this is currently hacky but it'll help us validate issues in submitted manifest files.

File: cmd/validate-krew-manifest/main.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
+// validate-krew-manifest makes sure a manifest file is valid.s
```
Comment (juanvallejo) [inline cmd/validate-krew-manifest/main.go]: nit typo at the end?

**Comment by juanvallejo** [comment]
@ahmetb trying this out locally. Currently running into the following error: 

```
$ go run cmd/validate-krew-manifest/main.go --manifest="~/gopaths/plugins/src/github.com/GoogleContainerTools/krew-index/plugins/change-ns.yaml"
I0204 13:36:44.913634   15279 main.go:57] structural validation OK
I0204 13:36:44.913756   15279 main.go:65] all spec.platform[] items are used
I0204 13:36:44.913850   15279 main.go:71] no overlapping spec.platform[].selector
I0204 13:36:44.913857   15279 main.go:75] installing spec.platform[0]
F0204 13:36:44.990111   15279 main.go:39] spec.platforms[0] failed to install: plugin install command failed: Error: unknown command "krew" for "kubectl"
        Run 'kubectl --help' for usage.
        unknown command "krew" for "kubectl"
        : exit status 1
exit status 255
```

**Comment by ahmetb** [comment]
@juanvallejo you need to install krew first. The validator runs `kubectl krew install` with `KREW_OS`, `KREW_ARCH` for each spec.platform[] and with a temp `KREW_ROOT` dir to see if installation goes through.

For example, the install failed to symlink "change-ns" on windows, because the `bin` field for windows was lacking the .exe suffix.

**Comment by juanvallejo** [comment]
@ahmetb yep thanks, had some path issues with kubectl. Appears to be working now

**Comment by ahmetb** [comment]
Ideally we'd take some of the validation here (like "has overlapping platform selectors", "all selectors map to at least one supported os/arch") and put it in `pkg/index#Validate()` method in the future.

File: cmd/validate-krew-manifest/main.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+			return errors.Errorf("multiple spec.platforms (at indexes %v) have overlapping selectors that select os=%s/arch=%s", matchIndex, os, arch)
+		}
+	}
+	return nil
+}
+
+// installPlatformSpec installs the p to a temporary location on disk to verify
+// by shelling out to external command.
+func installPlatformSpec(manifestFile string, p index.Platform) error {
+	goos, goarch := findAnyMatchingPlatform(p.Selector)
```
Comment (juanvallejo) [inline cmd/validate-krew-manifest/main.go]: @ahmetb testing this on linux. If I'm understanding correctly, this attempts to install manifest for all specified platforms. It currently fails on linux when installing for the `windows` platform: https://gist.github.com/juanvallejo/1ea80ea7406362ac9a78d517bbddb680

Comment (ahmetb) [reply cmd/validate-krew-manifest/main.go]: Change-ns manifest is actually broken. Windows spec should have “bin” field end in .exe.

**Comment by juanvallejo** [comment]
lgtm

**Pull request merged**